TVF Part 1. Pushdown to connector.

mohsaka opened the pull request:
This is the first part towards implementing Table functions in Presto. Please find the RFC at https://github.com/prestodb/rfcs/pull/39/files

In this PR, the Table function syntax is added to the Presto Parser, AST and RelationPlanner code is added. At this point, we only compile Table functions into TableScan of tables in the respective connector. There will be subsequent PRs to handle Table functions as operators (appearing mid-pipeline) and leaf operators.

Contains changes from https://github.com/prestodb/presto/pull/25032.

From commit 
[Support table function invocation in grammar and AST](https://github.com/prestodb/presto/pull/25032/commits/cb43449b812b8db8b29032240dba680a2c5ce9d5)

to commit
[Analyze table function arguments and cleanup](https://github.com/prestodb/presto/pull/25032/commits/6f7df408e8b07a61e9615017d82e45c4da96ced3)

This includes all of the TVF changes required to use table functions with pushdown to connector. This PR also includes as much refactoring changes as I could from future commits. 

## Description
<!---Describe your changes in detail-->
Allows table functions via pushdown to connector. Documentation will come at a later point after all of the code has been pushed in. Until then, this should be considered experimental. There is very little interaction with existing code.

Documentation from Trino can be followed as we wanted to match their existing API for portability.
https://trino.io/docs/current/functions/table.html

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->
TVF test cases added in `TestTableFunctionInvocation`.
Parser tests cases added in `TestSqlParser` and `TestSqlParserErrorHandling`.
JDBC failure test cases added in `TestJdbcDistributedQueries`.

Further testing is done in the complete PR.
## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

xin-zhang2 commented on presto-main-base/src/main/java/com/facebook/presto/metadata/TableFunctionRegistry.java:
```diff
@@ -0,0 +1,154 @@
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class TableFunctionRegistry
```
TableFunctionRegistry should be explicitly bound in VerifierModule and FunctionServerModule.
There is a failure in TestPrestoVerifierIntegrationSmokeTest.testVerifier caused by the missing binding.

xin-zhang2 commented on presto-spi/src/main/java/com/facebook/presto/spi/function/table/TableArgumentSpecification.java:
```diff
@@ -0,0 +1,103 @@
+            return this;
+        }
+
+        public Builder keepWhenEmpty()
```
It seems there is a potential issue here. `pruneWhenEmpty` is set to true when `rowSemantics()` is called, but it could be overwritten to false if `keepWhenEmpty()` is called afterward. While this is not an intended use pattern, we should still guard against it. Since `pruneWhenEmpty` is false by default, we might considering removing `keepWhenEmpty()`.

mohsaka replied:
I think its been covered in the constructor.

`checkArgument(!rowSemantics || pruneWhenEmpty, "Cannot set the KEEP WHEN EMPTY property for a table argument with row semantics");`

xin-zhang2 replied:
That's right. Let's keep this function as is.

aditi-pandit commented on presto-main-base/src/test/java/com/facebook/presto/connector/tvf/TestingTableFunctions.java:
```diff
@@ -0,0 +1,105 @@
+    public static class SimpleTableFunction
+            extends AbstractConnectorTableFunction
+    {
+        private static final String SCHEMA_NAME = "system";
```
Is there any issue if we use non-system schema here ?

mohsaka replied:
Tested with 'test' schema. Works. Left it at test for now.

aditi-pandit commented on presto-main-base/src/test/java/com/facebook/presto/connector/tvf/MockConnectorFactory.java:
```diff
@@ -0,0 +1,472 @@
+                        .map(MockConnectorColumnHandle::getType)
+                        .collect(toImmutableList());
+                Map<String, Integer> columnIndexes = getColumnIndexes(tableName);
+                /*
```
Can you remove this code if not needed ?

aditi-pandit commented on presto-main-base/src/test/java/com/facebook/presto/connector/tvf/TestingTableFunctions.java:
```diff
@@ -0,0 +1,105 @@
+        @Override
+        public TableFunctionAnalysis analyze(ConnectorSession session, ConnectorTransactionHandle transaction, Map<String, Argument> arguments)
+        {
+            ScalarArgument argument = (ScalarArgument) arguments.get("COLUMN");
```
What do you do with the IGNORED column ? Can we remove it if not required ?

aditi-pandit commented on presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java:
```diff
@@ -1312,4 +1342,52 @@ public int hashCode()
+                String functionName,
+                Map<String, Argument> arguments,
+                ConnectorTableFunctionHandle connectorTableFunctionHandle,
+                ConnectorTransactionHandle transactionHandle)
```
What does this do ?

aditi-pandit commented on presto-main-base/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -476,6 +477,28 @@ public PlanNode visitTableFinish(TableFinishNode node, RewriteContext<Void> cont
+        public PlanNode visitTableFunction(TableFunctionNode node, RewriteContext<Void> context)
+        {
+            // TODO rewrite sources, and tableArgumentProperties when we add support for input tables
+            /*
```
Remove this comment if not needed.

mohsaka replied:
Done

aditi-pandit commented on presto-main-base/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:
```diff
@@ -1308,6 +1309,32 @@ public Void visitLateralJoin(LateralJoinNode node, Void context)
+            return processChildren(node, context);
+        }
+
+        /*
```
Remove this code ?

mohsaka replied:
Done

aditi-pandit reviewed (commented):
Thanks @mohsaka and @xin-zhang2... Have a round of review comments.

xin-zhang2 commented on presto-spi/src/main/java/com/facebook/presto/spi/function/table/Descriptor.java:
```diff
@@ -0,0 +1,142 @@
+
+    public static class Field
+    {
+        private final Optional<String> name;
```
It's used in ReturnType to return anonymous columns.

mohsaka commented on presto-spi/src/main/java/com/facebook/presto/spi/function/table/AbstractConnectorTableFunction.java:
```diff
@@ -0,0 +1,68 @@
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractConnectorTableFunction
```
Done.

mohsaka commented on presto-spi/src/main/java/com/facebook/presto/spi/function/table/TableArgument.java:
```diff
@@ -0,0 +1,103 @@
+            return this;
+        }
+
+        public Builder partitionBy(List<String> partitionBy)
```
Done

mohsaka commented on presto-main-base/src/test/java/com/facebook/presto/connector/tvf/MockConnectorTableHandle.java:
```diff
@@ -0,0 +1,98 @@
+    @JsonCreator
+    public MockConnectorTableHandle(
+            @JsonProperty SchemaTableName tableName,
+            @JsonProperty("constraint") TupleDomain<ColumnHandle> constraint,
```
Done

mohsaka commented on presto-main-base/src/test/java/com/facebook/presto/connector/tvf/MockConnectorColumnHandle.java:
Done

mohsaka commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:
```diff
@@ -908,4 +909,17 @@ default String normalizeIdentifier(ConnectorSession session, String identifier)
+     * Attempt to push down the table function invocation into the connector.
+     * <p>
+     * Connectors can indicate whether they don't support table function invocation pushdown or that the action had no
+     * effect by returning {@link Optional#empty()}. Connectors should expect this method may be called multiple times.
```
Will add additional comment. Basically it means that we should expect the result to be consistent. For example, if there is some modification to the engine or class variables, it wouldn't be consistent.

mohsaka commented:
@aditi-pandit Addressed the comments. Please take a second look when you have a chance. The commit addressing comments will be merged into their respective commits once complete. Thanks!

mohsaka commented on presto-spi/src/main/java/com/facebook/presto/spi/function/CatalogSchemaFunctionName.java:
```diff
@@ -0,0 +1,95 @@
+public final class CatalogSchemaFunctionName
+{
+    private final String catalogName;
+    private final SchemaFunctionName schemaFunctionName;
```
It's mainly because we use SchemaFunctionName in the TableFunctionRegistry. So I guess 2 reasons.

We have the SchemaFunctionName class already that is widely used in TableFunctionRegistry. This is because, while in a Catalogs context, the catalog is no longer needed.
It allows for us to access this object directly rather than creating a SchemaFunctionName object each time.

pdabre12 commented on presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:
```diff
@@ -3472,4 +3483,161 @@ public void testSelectWithBeforeTimestamp()
     }
+
+    @Test
+    public void testTableFunctionInvocation()
```
Should we add some test cases for a descriptor with no fields e.g. `DESCRIPTOR()`? What is the expected behavior in this case?

mohsaka replied:
Thank you for the review. Descriptor() is not a valid SQL statement per the rule

```
descriptorArgument
    : DESCRIPTOR '(' descriptorField (',' descriptorField)* ')'
    | CAST '(' NULL AS DESCRIPTOR ')'
    ;
```

jaystarshot commented on presto-main-base/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java:
```diff
@@ -298,6 +301,44 @@ private RelationPlan addColumnMasks(Table table, RelationPlan plan, SqlPlannerCo
+    {
+        Analysis.TableFunctionInvocationAnalysis functionAnalysis = analysis.getTableFunctionAnalysis(node);
+
+        // TODO handle input relations:
```
its not clear yo me but looks like only scalar arguments work and table within table won't so can we fail gracefully for unsupported cases?

jaystarshot commented on presto-main-base/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteTableFunctionToTableScan.java:
```diff
@@ -0,0 +1,82 @@
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class RewriteTableFunctionToTableScan
```
Why do we need this rule and corresponding SPI calls? I think we can use Connector optimizers (https://prestodb.io/blog/2019/12/23/improve-presto-planner/) which are present in presto and not trino so we can potentially skip a lot of  additions to metadata?

jaystarshot replied:
cc: @tdcmeehan

jaystarshot commented on presto-main-base/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -409,6 +410,12 @@ public PlanWithProperties visitWindow(WindowNode node, PreferredProperties prefe
+        @Override
+        public PlanWithProperties visitTableFunction(TableFunctionNode node, PreferredProperties preferredProperties)
+        {
+            throw new UnsupportedOperationException("execution by operator is not yet implemented for table function " + node.getName());
```
Not sure we need this, by default it will throw unsupported

jaystarshot reviewed (commented):
I think the PR gets too big to review if we had a separate PR for parser only etc we could involve a lot more folks like @kaikalur  for the review/advice
